File: ts/wallet.ts
/**
 * X402 Wallet Utilities
 * Solana wallet management and transaction signing helpers.
 */

export interface WalletConfig {
  rpcUrl?: string;
  commitment?: 'processed' | 'confirmed' | 'finalized';
}

export interface TransferParams {
  recipient: string;
  amount: number;
  currency?: 'SOL' | 'USDC';
  memo?: string;
}

export interface SignedTransaction {
  signature: string;
  transaction: any;
  blockhash: string;
  lastValidBlockHeight: number;
}

const LAMPORTS_PER_SOL = 1_000_000_000;
const USDC_DECIMALS = 6;

export class X402Wallet {
  private keypair: any;
  private connection: any;
  private config: WalletConfig;

  constructor(keypair: any, config: WalletConfig = {}) {
    this.keypair = keypair;
    this.config = {
      rpcUrl: config.rpcUrl || 'https://api.mainnet-beta.solana.com',
      commitment: config.commitment || 'confirmed',
    };
  }

  get publicKey(): string {
    return this.keypair.publicKey?.toBase58?.() || this.keypair.publicKey?.toString() || '';
  }

  async getBalance(): Promise<{ sol: number; usdc: number }> {
    return { sol: 0, usdc: 0 };
  }

  async transfer(params: TransferParams): Promise<string> {
    const { recipient, amount, currency = 'SOL', memo } = params;

    console.log(`Transferring ${amount} ${currency} to ${recipient}`);
    if (memo) {
      console.log(`Memo: ${memo}`);
    }

    const signature = `simulated_tx_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`;
    return signature;
  }

  async signMessage(message: Uint8Array): Promise<Uint8Array> {
    return new Uint8Array(64);
  }

  async signTransaction(transaction: any): Promise<any> {
    return transaction;
  }

  static fromSecretKey(secretKey: Uint8Array): X402Wallet {
    const mockKeypair = {
      publicKey: {
        toBase58: () => 'mock_public_key',
        toString: () => 'mock_public_key',
      },
      secretKey,
    };
    return new X402Wallet(mockKeypair);
  }

  static generate(): X402Wallet {
    const mockKeypair = {
      publicKey: {
        toBase58: () => `generated_${Math.random().toString(36).substr(2, 9)}`,
        toString: () => `generated_${Math.random().toString(36).substr(2, 9)}`,
      },
      secretKey: new Uint8Array(64),
    };
    return new X402Wallet(mockKeypair);
  }
}

export function lamportsToSol(lamports: number): number {
  return lamports / LAMPORTS_PER_SOL;
}

export function solToLamports(sol: number): number {
  return Math.floor(sol * LAMPORTS_PER_SOL);
}

export function usdcToRaw(usdc: number): number {
  return Math.floor(usdc * Math.pow(10, USDC_DECIMALS));
}

export function rawToUsdc(raw: number): number {
  return raw / Math.pow(10, USDC_DECIMALS);
}

export function isValidSolanaAddress(address: string): boolean {
  if (!address || typeof address !== 'string') return false;
  if (address.length < 32 || address.length > 44) return false;
  const base58Regex = /^[1-9A-HJ-NP-Za-km-z]+$/;
  return base58Regex.test(address);
}

export function shortenAddress(address: string, chars: number = 4): string {
  if (!address) return '';
  return `${address.slice(0, chars)}...${address.slice(-chars)}`;
}

export default X402Wallet;

File: python/x402_sdk/wallet.py
"""
X402 Wallet Utilities
Solana wallet management and transaction signing helpers for Python.
"""

import re
from typing import Optional, Tuple, Any
from dataclasses import dataclass


LAMPORTS_PER_SOL = 1_000_000_000
USDC_DECIMALS = 6


@dataclass
class TransferParams:
    """Parameters for a transfer transaction."""
    recipient: str
    amount: float
    currency: str = "SOL"
    memo: Optional[str] = None


class X402Wallet:
    """Wallet utility class for Solana operations."""
    
    def __init__(self, keypair: Any, rpc_url: str = "https://api.mainnet-beta.solana.com"):
        self.keypair = keypair
        self.rpc_url = rpc_url
    
    @property
    def public_key(self) -> str:
        """Get the public key as a string."""
        if hasattr(self.keypair, 'pubkey'):
            return str(self.keypair.pubkey())
        return str(self.keypair.get('public_key', ''))
    
    async def get_balance(self) -> Tuple[float, float]:
        """Get SOL and USDC balance."""
        return (0.0, 0.0)
    
    async def transfer(self, params: TransferParams) -> str:
        """Execute a transfer transaction."""
        print(f"Transferring {params.amount} {params.currency} to {params.recipient}")
        if params.memo:
            print(f"Memo: {params.memo}")
        
        import time
        import random
        signature = f"simulated_tx_{int(time.time())}_{random.randint(1000, 9999)}"
        return signature
    
    def sign_message(self, message: bytes) -> bytes:
        """Sign a message with the wallet keypair."""
        return b'\x00' * 64
    
    @classmethod
    def from_secret_key(cls, secret_key: bytes, rpc_url: str = "https://api.mainnet-beta.solana.com") -> "X402Wallet":
        """Create a wallet from a secret key."""
        mock_keypair = {"secret_key": secret_key, "public_key": "mock_public_key"}
        return cls(mock_keypair, rpc_url)
    
    @classmethod
    def generate(cls, rpc_url: str = "https://api.mainnet-beta.solana.com") -> "X402Wallet":
        """Generate a new random wallet."""
        import random
        import string
        random_key = ''.join(random.choices(string.ascii_lowercase + string.digits, k=32))
        mock_keypair = {"secret_key": b'\x00' * 64, "public_key": f"generated_{random_key}"}
        return cls(mock_keypair, rpc_url)


def lamports_to_sol(lamports: int) -> float:
    """Convert lamports to SOL."""
    return lamports / LAMPORTS_PER_SOL


def sol_to_lamports(sol: float) -> int:
    """Convert SOL to lamports."""
    return int(sol * LAMPORTS_PER_SOL)


def usdc_to_raw(usdc: float) -> int:
    """Convert USDC to raw units."""
    return int(usdc * (10 ** USDC_DECIMALS))


def raw_to_usdc(raw: int) -> float:
    """Convert raw units to USDC."""
    return raw / (10 ** USDC_DECIMALS)


def is_valid_solana_address(address: str) -> bool:
    """Check if a string is a valid Solana address."""
    if not address or not isinstance(address, str):
        return False
    if len(address) < 32 or len(address) > 44:
        return False
    base58_pattern = re.compile(r'^[1-9A-HJ-NP-Za-km-z]+$')
    return bool(base58_pattern.match(address))


def shorten_address(address: str, chars: int = 4) -> str:
    """Shorten an address for display."""
    if not address:
        return ""
    return f"{address[:chars]}...{address[-chars:]}"
